<a href="https://colab.research.google.com/github/David-gutierrez-rodriguez/Ejercicioss/blob/main/Aplicaciones_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Durante una ronda de inspección por una reserva, dos guardabosques descubrieron
el cadáver de un jabalí. Un examen preliminar permitió concluir que el animal falleció
instantáneamente a causa de un balazo de una cazador furtivo. sospechando que el cazador debía regresar a recoger su presa, los guardas decidieron esperarlo escondidos no lejos del lugar del crimen. Al poco tiempo aparecieron dos sujetos y se dirigieron sin rodeos al jabalí. Al ser detenidos los desconocidos negaron rotundamente su participación en el delito. Aunque los guardabosques tenían pruebas indirectas
de su culpabilidad, para obtener una prueba fehaciente era necesario determinar el instante exacto en que el jabalí fue muerto.
Conforme a la ley de radiación de calor la velocidad de enfriamiento de un cuerpo en el medio ambiente es proporcionar a la diferencia. Entre la temperatura del cuerpo y la temperatura del medio

$$\frac{dx}{dt}=k(x-a)$$
Donde $x$ es la temperatura del cuerpo en el instante $t$, a es la temperatura del medio ambiente y $k$ es un factor positivo de proporcionalidad, $x_0$ es la temperatura del cuerpo en el instante $t = 0$.

a) Si en el instante en que fueron detenidos los sospechosos la temperatura $x$ del cuerpo del jabalí era igual a 31ºC y pasada una hora era igual a 29ºC, entonces, considerando que en el instante del disparo $x = 37$ºC y $a = 21$ºC, y tomando el instante del arresto como $t = 0$, determinar el momento del disparo.

b) En el caso cuando la temperatura del aire varía con el tiempo, la ley de enfriamiento del cuerpo se convierte en la ecuación diferencial lineal no homogénea.

$$\frac{dx}{dt}=kx-ka(t)$$

Donde $a(t)$ es la temperatura del aire en el instante $t$. Hagamos algunas conjeturas: en el instante de la detención de los sospechosos la temperatura del cuerpo del jabalí era igual a 30ºC; el día del acontecimiento la temperatura del aire descendió cada hora después del medio día en 1ºC y era igual a 0ºC en el instante del hallazgo del cadáver; finalmente, pongamos que al cabo de una hora de ser detectado el animal, su temperatura bajó hasta 25ºC. Tomando $t = 0$ como en el instante del disparo, asumiendo que entonces la temperatura corporal del jabalí era $x_0 = 37$ºC y denotando con $t^*$ el instante en que los guardabosques descubrieron el cadáver, obtenemos $a(t) = t^*-t$. Hallar de nuevo el momento en que murió el jabalí.

#Newton-Raphson

In [ ]:
import numpy as np
def f(x):
  return (5.621261 + 0.178953 * x) * np.e ** (0.178953 * t) - 4.36859
def g(x):
  return (1.184895 + 0.032024 * x) * np.e ** (0.178953 * x)
p_n = float(input("Ingresa la primera aporximación: "))
error = float(input("Ingresa el error de la aproximación: "))
e, c = 1, 0
print(f"n \t\t P_n \t\t\t Error")
while e > error:
  print(f"{c}\t {p_n}\t{e}")
  c += 1
  t = p_n
  if g(p_n) != 0:
    p_n -= f(p_n) / g(p_n)
  else:
    print("Divisón por cero")
    break
  e = abs(t - p_n)
if g(p_n) != 0:
  print(f"{c} \t {p_n} \t{abs(t - p_n)}")
  print(f"\n La raíz con un error de {error} es aproximadamente {p_n}")

Ingresa la primera aporximación: -1
Ingresa el error de la aproximación: 0.000001
n 		 P_n 			 Error
0	 -1.0	1
1	 -1.1887715996890686	0.18877159968906865
2	 -1.1925532801554792	0.0037816804664105863
3	 -1.192554762496123	1.4823406437880493e-06
4 	 -1.192554762497243 	1.119992987241858e-12

 La raíz con un error de 1e-06 es aproximadamente -1.192554762497243


# Solución a)

Tenemos la ecuación diferncial
$$\frac{dx}{dt}=k(x-a)$$
Usamos $t = 0$ como el tiempo de detención de los sospechosos

Resoleviendo por variables separables tenemos
$$ln\left(\frac{x-a}{x_0-a}\right) = -kt \quad (1)$$
Usando los datos del problema y sustituyendo en la ecuación anterior tenemos
$$k=ln\left(\frac{31-21}{29-21}\right) = 0.22314$$
Despejando de la ecuación (1) y sustituyendo el valor de $k$ y $x = 37$, obtenemos
$$t=\frac{1}{0.22314}ln\left(\frac{37-21}{31-21}\right) = -2.1063$$
Por lo tanto, el disparo al jabalí fue hecho 2 horas, 6 minutos y 23 segundos antes de que los sospechosos fueran detenidos

# Solución b)

En este caso, ya que la temmperatura del ambiente cambia respecto al tiempo tenemos la ecuación diferencial
$$\frac{dx}{dt} + kx = ka(t) \quad (1)$$
Tomando $t = 0$ como el momento del disparo, y $t^*$ como el instante en el que los guardabosques descubrieron el cadáver, obtenemos
$$a(t) = t^* - t$$
Haciendo la sustitución en (1) y resolviendo llegamos a
$$\frac{dx}{dt} + kx = k(t^* - t)$$
$$x = \left(37 - t^* - \frac{1}{k}\right)e^{-kt} + t^* - t + \frac{1}{k} \quad (2)$$
Teniendo en cuenta que $x = 30$ para $t^* = t$ , y que $x = 25$ para
$t = t^* + 1$, de la última expresión obtenemos las igualdades
$$\left(37 - t^* - \frac{1}{k}\right)e^{-kt^*} + \frac{1}{k} = 30$$
$$\left(37 - t^* - \frac{1}{k}\right)e^{-k(t^* + 1)} + \frac{1}{k} = 26$$
Luego entonces, tenemos:
$$\left(30 - \frac{1}{k}\right)e^{-k} - 26 + \frac{1}{k} = 0 \quad (3)$$
y la ecuación (2), después de tomar $x = 37$, toma la forma
$$(37k - 1 + kt)e^{kt} - 30k + 1 = 0 \quad (4)$$
Para resolver esto hacemos uso del método de Newton-Raphson, así, reescribimos a (2)
$$30k - 1 + (1 - 26k)e^k = 0$$
Luego definimos $f(k)$ y derivamos
$$f(k) = 30k - 1 + (1 - 26k)e^k$$
$$f'(k) = 30 - (25 + 26k)e^k$$
Notemos que $f'(k) = 0$ si $k = 0.0914595$
Además $f(0.1) \approx 0.2317$ y $f(1) \approx -38.95$, lo que nos dice que hay una raíz en [0.1,1]

Procedemos a usar el método de Newton con $x_0 = 0.5$

Usando un programa realizado en el curso, tenemos que $k = 0.178953$
Luego, susitiuímos el valor de $k$ en (4)
$$(6.621261 - 1 + 0.178953t)e^{0.178953t} - 5.36859 + 1 = 0$$
Para hallar el valor de $t$, procedemos de forma análoga
$$g(t) = (5.621261 + 0.178953t)e^{0.178953t} - 4.36859$$
$$g'(t) = (1.184895 + 0.032024t)e^{0.178953t}$$
Entonces $t = -1.192555$

Finalmente, podemos decir que el jabalí murió 1 hora, 11 minutos y 33 segundos antes de que lo encontraran
